File: sol/2147.py
class Solution:
    M = int(1e9) + 7

    def numberOfWays(self, corridor: str) -> int:
        plants = list()
        lastGroupEnd = -1
        for i, item in enumerate(corridor):
            if item == 'P':
                continue
            if lastGroupEnd is None:
                lastGroupEnd = i
            else:
                plantCount = i - lastGroupEnd - 1
                if plantCount != 0:
                    plants.append(plantCount)
                lastGroupEnd = None
        if lastGroupEnd is None or lastGroupEnd == -1:
            return 0
        s = 0
        if corridor[0] == 'P':
            s += 1
        res = 1
        for p in plants[s:]:
            res = (res * (p + 1)) % self.M
        return res

File: sol/2147/2147.cpp
struct Solution {
    int numberOfWays(const string &corridor) {
        constexpr int M = 1e9 + 7;
        vector<int> plants;
        int lastGroupEnd = -1, plantCount;
        for (int i = 0; i < corridor.size(); ++i){
            if (corridor[i] == 'P'){
                continue;
            }
            if (lastGroupEnd == -2){
                lastGroupEnd = i;
            } else {
                plantCount = i - lastGroupEnd - 1;
                if (plantCount != 0){
                    plants.push_back(plantCount);
                }
                lastGroupEnd = -2;
            }
        }
        if (lastGroupEnd < 0){
            return 0;
        }
        int s = 0;
        if (corridor[s] == 'P'){
            ++s;
        }
        long long int res = 1;
        for (int i = s; i < plants.size(); ++i){
            res = (res * (plants[i] + 1)) % M;
        }
        return res;
    }
};
